src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -160,35 +175,443 @@ private DeclaredTypeAssignment GetParameterType(ParameterDeclarationSyntax synta
+                    }
+                }
+            }
+
+            if (diagnostics.Any())
+            {
+                // foward any diagnostics gathered from parsing properties to the return type. normally, these diagnostics would be gathered by the SemanticDiagnosticVisitor (which would visit the properties of an ObjectType looking for errors).
+                // Errors hidden behind DeferredTypeReferences will unfortunately be dropped, as we can't call their type function without risking an infinite loop (in the case that a recursive object type has errors)
+                return ErrorType.Create(diagnostics.Concat(properties.Select(p => p.TypeReference).OfType<TypeSymbol>().SelectMany(e => e.GetDiagnostics())));
+            }
+
+            objectName.Remove(objectName.Length - 2, 1);
+            objectName.Append('}');
+
+            var parentSyntax = binder.GetParent(syntax);
+
+            var typeFlags = UnwrapUntilDecorable(syntax, HasSecureDecorator, TypeSymbolValidationFlags.IsSecure, TypeSymbolValidationFlags.Default);
+
+            (TypeSymbol? type, TypePropertyFlags flags) additionalProperties
+                = UnwrapUntilDecorable(syntax, HasSealedDecorator, (null, TypePropertyFlags.None), (LanguageConstants.Any, TypePropertyFlags.FallbackProperty));
+
+            return new ObjectType(objectName.ToString(), typeFlags, properties, additionalProperties.type, additionalProperties.flags);
+        }
+
+        private string GetPropertyTypeName(SyntaxBase typeSyntax, ITypeReference propertyType)
+        {
+            if (typeSyntax is TypeAccessSyntax typeAccess)
+            {
+                return typeAccess.Name.IdentifierName;
+            }
+
+            return propertyType switch
+            {
+                DeferredTypeReference deferred => "<deferred>",
+                _ => propertyType.Type.Name,
+            };
+        }
+
+        private string? TryGetTypePropertyDescription(ObjectTypePropertySyntax syntax)
+            => (TryGetDecorator(syntax.Decorators, SystemNamespaceType.BuiltInName, LanguageConstants.MetadataDescriptionPropertyName)
+                ?.Arguments.FirstOrDefault()?.Expression as StringSyntax)?.TryGetLiteralValue();
+
+        private T UnwrapUntilDecorable<T>(SyntaxBase syntax, Predicate<DecorableSyntax> condition, T valueIfTrue, T valueIfFalse) => binder.GetParent(syntax) switch
+        {
+            DecorableSyntax decorable when condition(decorable) => valueIfTrue,
+            ParenthesizedExpressionSyntax parenthesized => UnwrapUntilDecorable(parenthesized, condition, valueIfTrue, valueIfFalse),
+            TernaryOperationSyntax ternary when ternary.TrueExpression == syntax || ternary.FalseExpression == syntax
+                => UnwrapUntilDecorable(ternary, condition, valueIfTrue, valueIfFalse),
+            _ => valueIfFalse,
+        };
+
+        private bool HasSecureDecorator(DecorableSyntax syntax)
```

Comment:
Would be nice to have a common place to put these - extension method on `IBinder` or `DecorableSyntax`?

`public static DecoratorSyntax? TryGetSecureDecorator(this IBinder binder, DecorableSyntax syntax)`
`public static DecoratorSyntax? TryGetSecureDecorator(this DecorableSyntax syntax, IBinder binder)`